### PR TITLE
Extend the release script to tag and create the releases.

### PR DESCRIPTION
**Pull request by erikjohnston**
This splits the release script into three subcommands:

```
./scripts-dev/release.py prepare

# ... ask others to look at the changelog ...

./scripts-dev/release.py tag

# ... wait for asssets to build ...

./scripts-dev/release.py publish
```

Reviewable commit by commit.

### CODE REVIEW

File: scripts-dev/release.py [hunk 1]
```diff
@@ -240,6 +244,171 @@ def run():
...
+        raise click.ClickException("Uncommitted changes exist.")
+
+    current_version, _, _ = parse_version_from_module()
+    tag_name = f"v{current_version}"
+
+    if not click.confirm(f"Publish {tag_name}?", default=True):
+        return
+
+    if gh_token:
+        # Create a new draft release
```
Comment (richvdh) [inline scripts-dev/release.py:337]:
```suggestion
        # Publish the draft release
```

File: scripts-dev/release.py [hunk 2]
```diff
@@ -240,6 +244,171 @@ def run():
...
+        f"https://github.com/matrix-org/synapse/actions?query=branch%3A{tag_name}"
+    )
+
+    click.echo("Wait for release assets to be built")
+
+
+@cli.command()
+@click.option("--gh-token", envvar=["GH_TOKEN", "GITHUB_TOKEN"])
+def publish(gh_token: Optional[str]):
+    """Publish release and upload to PyPI."""
```
Comment (richvdh) [inline scripts-dev/release.py:319]: I can't help but feel these should be two separate commands, for clarity.

File: scripts-dev/release.py [hunk 3]
```diff
@@ -334,7 +334,7 @@ def publish(gh_token: Optional[str]):
         return
 
     if gh_token:
```
Comment (richvdh) [inline scripts-dev/release.py:336]: given that this now does nothing without a gh_token, I think it should error out rather than fail silently.

**Review by richvdh — COMMENTED** [review]
lgtm otherwise

File: scripts-dev/release.py [hunk 4]
```diff
@@ -240,6 +244,180 @@ def run():
...
+    for release in gh_repo.get_releases():
+        if release.title == tag_name:
+            break
+    else:
+        raise ClickException(f"Failed to find GitHub release for {tag_name}")
+
+    assert release.title == tag_name
+
+    if not release.draft:
+        if not click.confirm("Release already published. Continue?", default=True):
```
Comment (richvdh) [inline scripts-dev/release.py:348]: continue to... do nothing?

Comment (erikjohnston) [reply scripts-dev/release.py:348]: oh bah, less haste more speed or something

File: scripts-dev/release.py [hunk 5]
```diff
@@ -240,6 +244,180 @@ def run():
...
+        raise click.ClickException("Not in Synapse repo.")
+
+    if repo.is_dirty():
+        raise click.ClickException("Uncommitted changes exist.")
+
+    current_version, _, _ = parse_version_from_module()
+    tag_name = f"v{current_version}"
+
+    if not click.confirm(f"Publish {tag_name}?", default=True):
+        return
```
Comment (richvdh) [inline scripts-dev/release.py:334]: feel like this should be at the end, just before we actually publish it. but I'm nitpicking now.

**Review by richvdh — APPROVED** [review]
lgtm

**Pull request merged**